src/gd_avif.c

```diff
@@ -0,0 +1,684 @@
+	 Otherwise, return false.
+*/
+static avifBool isAvifError(avifResult result, const char *msg) {
+	if (result != AVIF_RESULT_OK) {
+		gd_error("avif error - %s: %s", msg, avifResultToString(result));
+		return AVIF_TRUE;
+	}
+
+	return AVIF_FALSE;
+}
+
+
+/*
+	<readfromCtx> implements the avifIOReadFunc interface by calling the relevant functions
+	in the gdIOCtx. Our logic is inspired by avifIOMemoryReaderRead() and avifIOFileReaderRead().
+	We don't know whether we're reading from a file or from memory. We don't have to know,
+	since we rely on the helper functions in the gdIOCtx.
+	We assume we've stashed the gdIOCtx in io->data, as we do in createAvifIOFromCtx().
+
+	We ignore readFlags, just as the avifIO*ReaderRead() functions do.
+
+	If there's a problem, this returns an avifResult error.
+	If things go well, return AVIF_RESULT_OK.
+	Of course these AVIF codes shouldn't be returned by any top-level GD function.
+*/
+static avifResult readFromCtx(avifIO *io, uint32_t readFlags, uint64_t offset, size_t size, avifROData *out)
+{
+	void *dataBuf = NULL;
+	gdIOCtx *ctx = (gdIOCtx *) io->data;
+
+	//TODO: if we set sizeHint, this will be more efficient.
+
+	if (offset > LONG_MAX || size < 0)
+		return AVIF_RESULT_IO_ERROR;
+
+	// Try to seek offset bytes forward. If we pass the end of the buffer, throw an error.
+	if (!ctx->seek(ctx, offset))
+		return AVIF_RESULT_IO_ERROR;
+
+	dataBuf = gdMalloc(size);
+	if (!dataBuf) {
+		gd_error("avif error - couldn't allocate memory");
+		return AVIF_RESULT_UNKNOWN_ERROR;
+	}
+
+	// Read the number of bytes requested.
+	// If getBuf() returns a negative value, that means there was an error.
+	int charsRead = ctx->getBuf(ctx, dataBuf, size);
+	if (charsRead < 0) {
+		gdFree(dataBuf);
+		return AVIF_RESULT_IO_ERROR;
+	}
+
+	out->data = dataBuf;
+	out->size = charsRead;
+	return charsRead == size ? AVIF_RESULT_OK : AVIF_RESULT_TRUNCATED_DATA;
+}
+
+// avif.h says this is optional, but it seemed easy to implement.
+static void destroyAvifIO(struct avifIO *io) {
+	avifFree(io);
+}
+
+/* Set up an avifIO object.
+	 The functions in the gdIOCtx struct may point either to a file or a memory buffer.
+	 To us, that's immaterial.
+	 Our task is simply to assign avifIO functions to the proper functions from gdIOCtx.
+	 The destroy function needs to destroy the avifIO object and anything else it uses.
+
+	 Returns NULL if memory for the object can't be allocated.
+*/
+
+//TODO: can we get sizeHint somehow?
+static avifIO *createAvifIOFromCtx(gdIOCtx *ctx) {
+	avifIO *io;
+
+	io = gdMalloc(sizeof(*io));
+	if (io == NULL)
+		return NULL;
+
+	io->read = readFromCtx;
+	io->write = NULL; // this function is currently unused; see avif.h
+	io->destroy = destroyAvifIO;
+	io->sizeHint = 0; // sadly, we don't get this information from the gdIOCtx.
+	io->persistent = AVIF_FALSE; // TODO: This seems like the safe thing to do, but I don't know for sure. This is less efficient, because it means AVIF will make copies of all buffers.
```

Comment:
should avoid inline comments like this.  they're very long and hard to read.  move it above the line and line wrap it.